signature/cose/envelope.go

```diff
@@ -0,0 +1,595 @@
+	// signingScheme
+	crit := []interface{}{headerLabelSigningScheme}
+	protected[headerLabelSigningScheme] = string(req.SigningScheme)
+
+	// signingTime/authenticSigningTime
+	signingTimeLabel, ok := signingSchemeTimeLabelMap[req.SigningScheme]
+	if !ok {
+		return &signature.InvalidSignRequestError{Msg: "signing scheme: require notary.x509 or notary.x509.signingAuthority"}
+	}
+	protected[signingTimeLabel] = req.SigningTime.Unix()
+	if signingTimeLabel == headerLabelAuthenticSigningTime {
+		crit = append(crit, headerLabelAuthenticSigningTime)
+	}
+
+	// expiry
+	if !req.Expiry.IsZero() {
+		crit = append(crit, headerLabelExpiry)
+		protected[headerLabelExpiry] = req.Expiry.Unix()
+	}
+
+	// extended attributes
+	for _, elm := range req.ExtendedSignedAttributes {
+		if _, ok := protected[elm.Key]; ok {
+			return &signature.InvalidSignRequestError{Msg: fmt.Sprintf("%q already exists in the protected header", elm.Key)}
+		}
+		if elm.Critical {
+			crit = append(crit, elm.Key)
+		}
+		protected[elm.Key] = elm.Value
+	}
+
+	// critical headers
+	protected[cose.HeaderLabelCritical] = crit
+
+	return nil
+}
+
+// generateUnprotectedHeaders creates Unprotected Headers of the COSE envelope
+// during Sign process.
+func generateUnprotectedHeaders(req *signature.SignRequest, signer signer, unprotected cose.UnprotectedHeader) {
+	// signing agent
+	unprotected[headerLabelSigningAgent] = req.SigningAgent
+
+	// certChain
+	certs := signer.CertificateChain()
+	certChain := make([]interface{}, len(certs))
+	for i, c := range certs {
+		certChain[i] = c.Raw
+	}
+	unprotected[cose.HeaderLabelX5Chain] = certChain
+}
+
+// parseProtectedHeaders parses COSE envelope's protected headers and populates signature.SignerInfo
+func parseProtectedHeaders(protected cose.ProtectedHeader, signerInfo *signature.SignerInfo) error {
+	// validate critical headers and return extendedAttributeKeys
+	extendedAttributeKeys, err := validateCritHeaders(protected)
+	if err != nil {
+		return err
+	}
+
+	// populate signerInfo.SignatureAlgorithm
+	alg, err := protected.Algorithm()
+	if err != nil {
+		return err
+	}
+	sigAlg, ok := coseAlgSignatureAlgMap[alg]
+	if !ok {
+		return &signature.InvalidSignatureError{Msg: "signature algorithm not supported: " + strconv.Itoa(int(alg))}
+	}
+	signerInfo.SignatureAlgorithm = sigAlg
+
+	// populate signerInfo.SignedAttributes.SigningScheme
+	signingSchemeString, ok := protected[headerLabelSigningScheme].(string)
+	if !ok {
+		return &signature.InvalidSignatureError{Msg: "malformed signingScheme"}
+	}
+	signingScheme := signature.SigningScheme(signingSchemeString)
+	signerInfo.SignedAttributes.SigningScheme = signingScheme
+
+	// populate signerInfo.SignedAttributes.SigningTime
+	signingTimeLabel, ok := signingSchemeTimeLabelMap[signingScheme]
+	if !ok {
+		return &signature.InvalidSignatureError{Msg: "unsupported signingScheme: " + signingSchemeString}
+	}
+	signingTime, ok := protected[signingTimeLabel].(int64)
```

Comment:
Do we need to check `signingTimeLabel` exists in the `protected` first?

Reply:
signingTime, ok := protected[signingTimeLabel].(int64) should check both existence and type assertion right?

Reply:
> signingTime, ok := protected[signingTimeLabel].(int64) should check both existence and type assertion right?

Yes. Just we don't know which error it is. But if we don't need the detail message, it's ok 😄